File: data/src/main/java/org/cryptomator/data/util/TransferredBytesAwareOutputStream.java
package org.cryptomator.data.util;

import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.OutputStream;

public abstract class TransferredBytesAwareOutputStream extends OutputStream {

	private final OutputStream out;

	private long transferred;

	public TransferredBytesAwareOutputStream(OutputStream out) {
		this.out = out;
	}

	@Override
	public void write(byte @NotNull [] b) throws IOException {
		out.write(b);
		transferred += b.length;
		bytesTransferred(transferred);
	}

	@Override
	public void write(byte @NotNull [] b, int off, int len) throws IOException {
		out.write(b, off, len);
		transferred += len;
		bytesTransferred(transferred);
	}

	@Override
	public void write(int i) throws IOException {
		out.write(i);
		bytesTransferred(++transferred);
	}

	@Override
	public void close() throws IOException {
		out.close();
	}

	@Override
	public void flush() throws IOException {
		out.flush();
	}

	public abstract void bytesTransferred(long transferred);

}

File: data/src/main/java/org/cryptomator/data/util/TransferredBytesAwareOutputStream.kt
package org.cryptomator.data.util

import java.io.IOException
import java.io.OutputStream

abstract class TransferredBytesAwareOutputStream(private val outputStream: OutputStream) : OutputStream() {

	private var transferred: Long = 0

	@Throws(IOException::class)
	override fun write(b: ByteArray) {
		outputStream.write(b)
		transferred += b.size.toLong()
		bytesTransferred(transferred)
	}

	@Throws(IOException::class)
	override fun write(b: ByteArray, off: Int, len: Int) {
		outputStream.write(b, off, len)
		transferred += len.toLong()
		bytesTransferred(transferred)
	}

	@Throws(IOException::class)
	override fun write(i: Int) {
		outputStream.write(i)
		bytesTransferred(++transferred)
	}

	@Throws(IOException::class)
	override fun close() {
		outputStream.close()
	}

	@Throws(IOException::class)
	override fun flush() {
		outputStream.flush()
	}

	abstract fun bytesTransferred(transferred: Long)
}
